flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -575,40 +582,62 @@ class ScalarExprLowering {
       if (!inInitializer && sym->Rank() > 0)
         TODO(loc, "array component in structure constructor");
 
-      auto val = genval(expr.value());
-      auto valTy = fir::getBase(val).getType();
-      if (fir::isa_ref_type(valTy)) {
-        assert(!inInitializer && "expecting a constant value");
-        auto fldType = fir::dyn_cast_ptrEleTy(valTy);
-        if (fir::isa_char(fldType)) {
-          auto fldCharTy = fldType.cast<fir::CharacterType>();
-          if (!fldCharTy.hasConstantLen())
-            TODO(loc, "CHARACTER LEN type parameter not constant");
-          auto fldRefTy = builder.getRefType(fldType);
-          auto lenVal = builder.createIntegerConstant(loc, builder.getI64Type(),
-                                                      fldCharTy.getLen());
-          mlir::Value from = builder.create<fir::CoordinateOp>(
-              loc, fldRefTy, fir::getBase(val), field);
-          ExtValue fromPtr{fir::CharBoxValue{from, lenVal}};
-          mlir::Value to = builder.create<fir::CoordinateOp>(
-              loc, fldRefTy, fir::getBase(res), field);
-          ExtValue toPtr{fir::CharBoxValue{to, lenVal}};
-          Fortran::lower::CharacterExprHelper{builder, loc}.createAssign(
-              toPtr, fromPtr);
-          continue;
+      mlir::Value val;
+      if (isBuiltinCPtr(sym)) {
+        // Builtin c_ptr and c_funptr have special handling because initial
+        // value are handled for them as an extension.
+        auto addr = inInitializer ? Fortran::lower::genExtAddrInInitializer(
+                                        converter, loc, expr.value())
+                                  : genExtAddr(expr.value());
+        auto baseAddr = fir::getBase(addr);
+        auto undef = builder.create<fir::UndefOp>(loc, componentTy);
+        auto cPtrRecTy = componentTy.dyn_cast<fir::RecordType>();
+        assert(cPtrRecTy);
+        auto addrFieldName = "__address";
```

Comment:
This looks like a magic constant. Is it part of an interface?

Reply:
It's only defined in __fortran_type_info.f90  [here ](https://github.com/flang-compiler/f18-llvm-project/blob/c1db35f0c232a8672d44f2531d178d4da35b5b3c/flang/module/__fortran_type_info.f90#L21), so not really accessible here. I agree it is a bit magic here. I am not sure how to share this in a best way (all I can see for now would be some kind of header files with macros shared in the two files).
We could segregate those runtime magic strings somewhere in the meantime in lowering though.

Reply:
Putting these bits of magic in a `Voodoo.h` header file with some comments seems like a very good idea.

Reply:
Moved the magic string into a new `lib/Lower/BuitltinModules.h`